1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,611 @@
+}
+```
+
+#### Background operator changes
+
+In bash, background operators are lower precedence than chain separators
+(`&` has the same precedence as `;`).
+In PowerShell, background operators are higher precedence.
+
+Making chain operators apply to any statement would lead to the question
+of the precedence of the background operator.
+
+The example:
+
+```powershell
+cmd1 && cmd2 &
+```
+
+Could either be `[[cmd1 && cmd2] &]` or `[cmd1 && [cmd2 &]]`.
+
+Having chain operators apply to statements would mean the second case
+is the correct one, unless changes are made to the background operator precedence (which would also allow `if ($condition) { cmd1 } &)` for example).
+
+The use of this is possibly less than the use of being able to background
+an entire pipeline chain; `cmd1 & && cmd2` would be equivalent to
+`cmd1 &; cmd2` since a backgrounded pipeline
+will never fail in the invoking context.
+
+Such a change to the background operator precedence
+would cause ambiguity with uses like `return cmd1 &` and `throw cmd1 &`,
+where the background operator would currently apply to the pipeline under the keyword.
+To not break PowerShell's existing semantics,
+special behaviour would need to be defined for `return $expr &`.
+
+#### Reasons against
+
+- Pipelines have an established concept of "success" compared to statements
+- Background operators become less useful with respect to chains unless their
+  syntax is changed in a significant way
+
+### Allowing control flow statements at the end of chains
+
+A compromise to the above is to only allow "control flow statements"
+at the end of chains.
+
+For example:
+
+```powershell
+cmd1 || throw "cmd1 failed"
+```
+
+```powershell
+function Invoke-Command
+{
+    cmd1 && return
+
+    cmd2
+}
+```
+
+```powershell
+foreach ($v in 1..100)
+{
+    cmd1 $v && break
+}
+```
+
+These would only be allowed *at the end* of chains because:
+
+- Control flow makes any invocation after the statement unreachable
+- `return` and `throw` allow pipelines as subordinate expressions,
+   meaning it would be grammatically impossible to use those statements mid-chain.
+   See below for more details.
+
+#### Reasons against
+
+- `throw` and `return` currently allow a subordinate pipeline:
+
+   ```powershell
+   throw "Bad"
+   ```
+
+   ```powershell
+   throw 1,2,3 | Write-Output
+   ```
+
+   ```powershell
+   throw "Useless" & # Throws a background job...
+   ```
+
+   If we keep the principle that anywhere a pipeline is allowed,
+   a pipeline chain is now allowed, then constructions like this are possible:
+
+   ```powershell
+   cmd1 || throw "Bad" && cmd3
+   ```
+
+   However the grouping is this:
+
+   ```none
+   [cmd1 || throw ["Bad" && cmd3]]
+   ```
+
+   This could lead to confusing semantic corner cases
+
+- The compromise nature of this approach means the PowerShell Language
+  becomes more complicated and arguably less consistent,
+  both conceptually and in terms of maintenance.

```

Comment:
Might be better to have subsequent pipeline chains require the use of a subexpression or something, rather than implicitly allowed here. This does muddy the waters a bit, but if it is deliberately prevented here, it would be more or less OK, I feel.

Reply:
Could you expand a bit more on the subexpression requirement, maybe an example of what you're thinking?

Reply:
So rather than just allowing pipeline chains to follow something like `throw` or `return` or whatever...

Attempting to do so without enclosing it should be a parse error, imo. But you should still be able to do it by properly enclosing it:
```ps
cmd1 || throw ("Bad" && cmd3)
```

Reply:
Do you mean `cmd1 || throw "Bad" && cmd_cleanup` ?

Reply:
That's what I'm saying I don't think should be permitted.

It _looks_ like the `&&` should never trigger there, because of the throw. But in fact it will because after the `throw` it's a separate pipeline chain.

I feel this should be required to be enclosed, because otherwise it's very unclear what will happen  in these situations.

Reply:
@vexx32: That's a nice way to allow flow control with pipeline chains while dodging semantic confusion. I like it.

Reply:
I was going to write something about this being complicated in terms of the grammar because of needing to support simple pipelines without parens, but it might not be as complicated as I first thought.

Essentially `throw <expr>` currently allows a pipeline statement where `<expr>` is:

- `throw "Bad"`
- `throw 1,2,3 | % { "Bad $_" }`

The original proposal in the RFC is to just allow pipeline chains wherever a pipeline is, but we can modify that so that `throw` still only allows a pipeline (no chain). A pipeline can optionally start with a pipeline expression, like `(cmd1 && cmd2)`, or a subexpression, like `$(if (...) { ... })` (that's always been the case), so that works out.

I think I can deal with that, even though it makes the grammar more complicated. (The reason I want to avoid that is because it's a common trap that's afflicted everything from [PL/I](https://en.wikipedia.org/wiki/PL/I#Criticisms) to [Python](https://www.reddit.com/r/programming/comments/8vsy0u/controversial_assignment_expressions_proposal/) -- keeping a grammar consistent enough that you can reason about it in your head without being reminded of things usually means you aren't engineering awful pitfalls into your language that you'll never be able to get rid of.)